zulip/zulip/__init__.py

```diff
@@ -697,24 +697,41 @@ def call_endpoint(
         method: str = "POST",
         request: Optional[Dict[str, Any]] = None,
         longpolling: bool = False,
+        retry_on_rate_limit_error: bool = False,
         files: Optional[List[IO[Any]]] = None,
         timeout: Optional[float] = None,
     ) -> Dict[str, Any]:
+        secs: float
+
         if request is None:
             request = dict()
         marshalled_request = {}
         for (k, v) in request.items():
             if v is not None:
                 marshalled_request[k] = v
         versioned_url = API_VERSTRING + (url if url is not None else "")
-        return self.do_api_query(
-            marshalled_request,
-            versioned_url,
-            method=method,
-            longpolling=longpolling,
-            files=files,
-            timeout=timeout,
-        )
+
+        while True:
+            result = self.do_api_query(
+                marshalled_request,
+                versioned_url,
+                method=method,
+                longpolling=longpolling,
+                files=files,
+                timeout=timeout,
+            )
+            if not retry_on_rate_limit_error or result["result"] == "success":
+                break
+            elif "code" in result and result["code"] == "RATE_LIMIT_HIT":
+                secs = result["retry-after"]
+            elif "X-RateLimit-Reset" in result:
+                secs = float(result["X-RateLimit-Reset"])
```

Comment:
So, I think the `retry-after` property is there since the beginning of time for these errors; just not `code`. So I think you could do 
```
if 'retry-after' not result:
    break
 
wait_before_retry_secs = result["retry-after"]
logger.warning(...)
time.sleep(...)